### ISSUE TITLE
Some improvement ideas for rt.sh

### ISSUE THREAD

**Issue by uturuncoglu** [ufs-community/ufs-weather-model#1837]
## Description
At this point, we are working on developing new [ufs-coastal](https://github.com/oceanmodeling/ufs-coastal/tree/feature/coastal_app) model and it would be nice to have more flexibility in exiting RT system that would allow users to overwrite `DISKNM` and `dprefix` via command line arguments since the components and regression tests are not part of the ufs-weather-model yet (I am not sure it will be in the future). For support to skip compile step after initial run of RTs will also boost development of new RTs. I could create a very small PR that bring the capabilities if there is an interest about it.

## Solution
Update rt.sh to allow those new features along with command line arguments.

## Alternatives
N/A

## Related to
N/A

**Comment by ulmononian** [comment]
i support these suggestions!!!

**Comment by ulmononian** [comment]
can you also add option to retain rocoto .db and .xml files after the RTs are run w/ rocoto? rocotostat output is much easier to decipher than the other rt logs when there are 130 tests for intel alone.

**Comment by mkavulich** [comment]
For what it's worth, I have already undertaken a similar effort on the NCAR fork as part of my efforts to get automated regression tests working again for CCPP...it was not a pleasant experience. It took me days to get the system working on Hera, and I have yet to get it working on Cheyenne due to some mysterious module environment issues. Here is the PR with the changes if that is helpful to whoever is working on this tasks: https://github.com/NCAR/ufs-weather-model/pull/101/files#diff-998e7472e54700d07abca0f198f31f64ecac11bed05f40b61756d112f826c7e5

I would love to see the regression testing improved, but in my opinion (for what that's worth) it would be better to re-write rt.sh from scratch in python (or some other modern language) rather than trying to hack these things in. The current chaotic spiderweb of hard-coded bash is a nightmare to try to modify in any way.

**Comment by uturuncoglu** [comment]
@ulmononian Sure. Let me work on that. 
@mkavulich It is great to have your commit. I think it is better to push those changes to UFS authoritative repository rather than keeping in the fork. So, other applications could also benefit from it like coastal app. Do you want me to work on it by using your PR as reference? Since UFS is also used for operational purposes, it might require lots of restrictions that we don't aware of. So, maybe this is the reason of written using shell script. Since, it does not require any dependency etc. it would be nice to have it with this way. Anyway, thanks for sharing your work. It is good to know that others are also thinking about these improvements.

**Comment by mkavulich** [comment]
@uturuncoglu The NCAR fork of the ufs-weather-model is continuously synced with the ufs-community fork, but we have maintained our own script (rt.ncar.sh) specifically because so many things (output location, baseline location, etc.) are hard-coded to the specific way that rt.sh is run for the ufs-community repository testing. If we can meet in the middle to make the script configurable so that it can meet the needs of both the weather model team and the CCPP team (and everyone else!) then I think that would be very beneficial.

Please feel free to adapt my changes linked above to the main rt.sh; if you ping me when the PR is open I can comment on any changes that might be needed to be included in order for us to retire our own script in favor of the main rt.sh. Or feel free to reach out to me via email (kavulich@ucar.edu) if you'd like to discuss a different strategy.

**Comment by BrianCurtis-NOAA** [comment]
I completely agree with all these suggestions. I think a python version of the regression testing system would be best, and can be suited to fit all these needs and much more. Before I joined EMC there was a python version but only one person knew the code and left before passing along his knowledge of the code, so they reverted to a shell scripting system.

I think in todays world, python isn't what it was then and proper coding standards would not leave us in the same situation as before.

**Comment by mkavulich** [comment]
@uturuncoglu I am curious if you have been working on this. I have put some work into unifying the two sets of regression testing scripts (still in the same branch linked above), but before I go too much further I want to make sure that I'm not diverging from any concurrent work.

**Comment by uturuncoglu** [comment]
@mkavulich Sorry, I could not do anything due to the other staff. It is still in my radar but not sure when I could make it. This is actually required for the UFS Coastal but I am currently porting the model components and its priority is low when it is compared with other staff.

**Comment by mkavulich** [comment]
@uturuncoglu No worries, I was only commenting here to make sure I wasn't doing any duplicate/divergent work! I have opened a new issue (#1901) to track my work.

**Comment by uturuncoglu** [comment]
@junwang-noaa @ulmononian @mkavulich In addition to my previous comments, the existing RT system assumes that the RTs can be only run on certain platforms such as Tier 1 machines or cloud. It would be also nice to have flexibility to run exiting RTs on custom machines such as unsupported clusters, Linux desktops or even Docker container (I could able to use spack-stack to build configurations under Docker container) with possibility of not having jobs schedular/ecflow or rocoto. At this point this requires extensive work in user side and it is not trivial and some capabilities are missing. So, it would be nice to support those cases too and also able to run RTs since the model could be ported easily with spack-stack. I think this improves the portability and usability of the model. Since input files are available through the S3 bucket, the user could get those files (or a subset of it) and try to run RTs (probably computationally cheap ones) on their local machine. I am not sure but maybe it might be nice to setup call with the developers to discuss the current limitations of the RT framework and possible design approaches to overcome those issues. Anyway, let me know what you think.

**Comment by gspetro-NOAA** [comment]
@uturuncoglu Just checking in to see if you are still planning on working on this issue!

**Comment by uturuncoglu** [comment]
@gspetro-NOAA Thanks for checking. At this point my answer is no. I think we could close this and open new one if we need in the future.

**Comment by gspetro-NOAA** [comment]
If/when work is underway to update `rt.sh`, developers can open a new issue documenting planned changes, which _could_ include the changes proposed in this issue. 
Note that a python-based RT framework is available through `tests-dev`, and EPIC developers will open new issues as they work to improve/expand that.
